### PR TITLE
Memory management

### LINKED ISSUE #40: Remap kernel
The kernel needs to be remapped such that a kernel stack overflow doesn't corrupt the page tables, as it is located above the page tables as per Phil Opp's tutorial before he remaps the kernel.

### PR DESCRIPTION
**Pull request by Restioson**
Putting this here for tracking and visibility.

To do list:
- [x] PMM
- [x] Paging
- [x] Huge page support
- [x] Kernel heap
- [x] Two stage PMM init
- [ ] Unmap parent/grandparent page tables

Fixes #40.

### CODE REVIEW

**Comment by Restioson** [comment]
@gegy1000 do you think we need to implement higher half & memory protections *now*, or should we do this when userspace is implemented (this is the reason we will do it).

**Comment by Restioson** [comment]
_Or is it just because you want heap now?_ Just kidding, I agree.

File: kernel/src/lang.rs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 //! Lang items
 
+use core::f32;
```
Comment (gegy1000) [inline kernel/src/lang.rs]: Unused import here.

File: kernel/src/interrupts/exceptions.rs [hunk 2]
```diff
@@ -19,7 +19,12 @@ pub extern "x86-interrupt" fn out_of_bounds(stack_frame: &mut ExceptionStackFram
 }
 
 pub extern "x86-interrupt" fn invalid_opcode(stack_frame: &mut ExceptionStackFrame) {
-    panic!("cpuex: invalid opcode\n{:#?}", stack_frame);
+    panic!(
+        "cpuex: invalid opcode \n{:#?}\n => note: qword at {:?} is 0x{:x}",
```
[the remark this answers is not in the document]
Comment (Restioson) [reply kernel/src/interrupts/exceptions.rs]: We don't, but we just print out 4 bytes there to make it easier to debug.

File: kernel/src/memory/bootstrap_heap.rs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+#[derive(Debug)]
+pub struct BootstrapAllocator<T> {
+    start_addr: usize,
+    objects8: Mutex<[Objects8; OBJECTS8_NUMBER]>,
+    _phantom: PhantomData<T>,
+}
+
+/// A bitmap consisting of 8 objects
+#[derive(Debug, Copy, Clone)]
+struct Objects8(u8);
```
Comment (gegy1000) [inline kernel/src/memory/bootstrap_heap.rs]: Why not directly use a u8? The name Objects8 is also quite unclear as to what it is used for.

Comment (Restioson) [reply kernel/src/memory/bootstrap_heap.rs]: I mean, the doc explains it, doesn't it? I would have called it 8Objects but that isn't allowed (begins with a number). I use a newtype for clarity.

File: kernel/src/memory/bootstrap_heap.rs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    /// Set a block to used or not at an index
+    #[inline]
+    fn set_used(&self, index: usize, used: bool) {
+        let objects8_index = index / 8;
+        let bit = index % 8;
+        self.objects8.lock()[objects8_index].0.set_bit(bit, used);
+    }
+
+    /// Allocate an object of zeroes and return the address if there is space
+    unsafe fn allocate_zeroed<'a>(&'a self) -> Option<BootstrapHeapBox<'a, T>> {
```
Comment (gegy1000) [inline kernel/src/memory/bootstrap_heap.rs]: Why is the buddy allocator not used here?

Comment (Restioson) [reply kernel/src/memory/bootstrap_heap.rs]: Takes a lot of space to store and is relatively complex four our purposes. The bootstrap heap is used once (at boot), so it's simpler to just do a bitmap.

File: kernel/src/memory/heap.rs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        let order = util::ceil(order) as u8;
+
+        if order > MAX_ORDER {
+            0 as *mut _
+        } else {
+            if let Some(ptr) = tree.allocate(order) {
+                let ptr = (ptr as usize + HEAP_START) as *mut _;
+
+                // Map ptr page if it isn't already
+                let page_tables = PAGE_TABLES; // What the fuck... This is needed
```
Comment (gegy1000) [inline kernel/src/memory/heap.rs]: Make sure TODO before merge. :P

Comment (Restioson) [reply kernel/src/memory/heap.rs]: Which TODO?

File: kernel/src/memory/physical_allocator.rs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    {
+        let mut flat_blocks: BootstrapHeapBox<[Block; BLOCKS_IN_TREE]> = unsafe {
+            super::bootstrap_heap::BOOTSTRAP_HEAP.allocate_zeroed()
+            .expect("No heap memory available!")
+        };
+        Self::new(usable, flat_blocks)
+    }
+}
+
+pub struct PhysicalAllocator<'a> {
```
Comment (gegy1000) [inline kernel/src/memory/physical_allocator.rs]: Might be better to make a separate struct for testing.

File: kernel/src/interrupts/exceptions.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (gegy1000) [reply kernel/src/interrupts/exceptions.rs]: The log states "qword at", should this not be removed?

File: kernel/src/memory/bootstrap_heap.rs (hunk 3 above)
Comment (gegy1000) [reply kernel/src/memory/bootstrap_heap.rs]: 8Objects doesn't seem any more clear to me. What about rather abstracting access to the bitmap for clarity?

File: kernel/src/memory/bootstrap_heap.rs (hunk 4 above)
Comment (gegy1000) [reply kernel/src/memory/bootstrap_heap.rs]: Ah. 👍

File: kernel/src/interrupts/exceptions.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Restioson) [reply kernel/src/interrupts/exceptions.rs]: It _is_  the qword at that address (qword = 8 bytes) (I got it mixed up when I said 4 bytes  earlier, we do really print 8). It's just a question of if the bytes are relevant or not. Often one byte is not enough, as it is just an instruction prefix.

File: kernel/src/memory/physical_allocator.rs (hunk 6 above)
Comment (Restioson) [reply kernel/src/memory/physical_allocator.rs]: Why? It's just simpler to use cfg's, and we don't have to keep synchronising the testing and the non testing. It's also less code.

File: kernel/src/memory/heap.rs (hunk 5 above)
Comment (gegy1000) [reply kernel/src/memory/heap.rs]: "Wtf" code is not really ideal.

Comment (Restioson) [reply kernel/src/memory/heap.rs]: _Yeah, but it's true..._ Okay, fine.

**Comment by gegy1000** [comment]
This looks great! Awesome work. :)

**Comment by Restioson** [comment]
Thanks! I will fix up the stuff later.

File: kernel/src/memory/bootstrap_heap.rs (hunk 3 above)
Comment (Restioson) [reply kernel/src/memory/bootstrap_heap.rs]: Okay, I'll make it into just a plain u8.

**Comment by Restioson** [comment]
The last TODOs left are about making the unmap function free the parent tables if they are free. Unfortunately, doing this would be very slow, so unless another solution presents itself it isn't really worth doing it now (not *much* is gained, anyway). Can we merge this @gegy1000 ?

File: kernel/src/memory/physical_allocator.rs (hunk 6 above)
Comment (gegy1000) [reply kernel/src/memory/physical_allocator.rs]: Fair, though I'd imagine it'd introduce better readability. For example, I got confused why the `tree` field was being declared twice.

Comment (Restioson) [reply kernel/src/memory/physical_allocator.rs]: In this case, imo the complications outweigh the gains. Any way you do it, something will be declared twice.

**Comment by gegy1000** [comment]
What issues could come up through the lack of implementation for this TODO?

File: kernel/src/memory/physical_allocator.rs (hunk 6 above)
Comment (gegy1000) [reply kernel/src/memory/physical_allocator.rs]: It's less about things being declared twice than it is about separation of concerns. Fair enough, though. Either works.

**Comment by Restioson** [comment]
Uh... We can waste a bit of space potentially, but in the long run the chances are that eventually the pages will be remapped. Even if they aren't the bit of space shouldn't matter tooo much. It's 4kib per page table, so if we can free a whole 4kib-level page table we gain about 4kib back from the 2mib that it maps.

File: kernel/src/memory/bootstrap_heap.rs [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+use std::boxed::Box;
+use core::{mem, ptr::{self, Unique}};
+use core::marker::PhantomData;
+use core::ops::{Deref, DerefMut};
+use spin::{Once, Mutex};
+use bit_field::BitField;
+use super::{physical_allocator::BLOCKS_IN_TREE, buddy_allocator::Block};
+
+/// How many 8 heap objects to allocate at maximum.
+// TODO two stage bootstrap -- allocate 1gib on the bootstrap heap and the rest on kernel heap
```
Comment (gegy1000) [inline kernel/src/memory/bootstrap_heap.rs]: TODO

File: kernel/src/memory/paging.rs [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+        let p1 = p2.next_page_table_mut(page.p2_index());
+
+        if let Some(p1) = p1 {
+            // 4kib page
+
+            let frame = p1[page.p1_index()].physical_address().unwrap();
+            p1[page.p1_index()].set_unused();
+
+            // TODO free p1/p2/p3 tables if they are empty
+            PHYSICAL_ALLOCATOR.deallocate(frame.0 as *const _, 0);
```
Comment (gegy1000) [inline kernel/src/memory/paging.rs]: TODO

File: kernel/src/memory/paging.rs [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+        } else {
+            // Huge 2mib page
+
+            let frame = p2[page.p2_index()].physical_address().unwrap();
+            p2[page.p2_index()].set_unused();
+
+            // TODO free p2/p3 tables if they are empty
+            PHYSICAL_ALLOCATOR.deallocate(frame.0 as *const _, 0);
```
Comment (gegy1000) [inline kernel/src/memory/paging.rs]: TODO

**Comment by Restioson** [comment]
The latest commit Works On My Machine:tm:. Could you update rust on the jenkins? Same thing happened to me, seems to be fixed on yesterday's nightly.

**Comment by Restioson** [comment]
Btw, I *think* we can make unmapping parents/grandparents O(1) -- we can store the amount of child tables present in the free bits (52-62) in the first entry of a given table. I'll do this when I have time.

**Comment by Restioson** [comment]
I think we could do the unmapping later. It'd be good to get this in now, and unmapping the page tables when they are free doesn't give you *that* much.

**Comment by gegy1000** [comment]
@Restioson I think we should at least structure it in such a way that there are not TODOs lying around.

**Comment by Restioson** [comment]
There are only two TODOs in relevant places though (that I can see).

**Comment by gegy1000** [comment]
Woot! 👏

**Pull request merged**